### ISSUE TITLE
Compatibility with Express syntax

### ISSUE THREAD

**Issue by ramlmn** [lukeed/polka#22]
Currently the middleware `serve-static` as in the examples is implemented as
``` js
polka()
  .get('/', serve)
  .use('assets', serve)
  .get('/health', (req, res) => {
    res.end('OK');
  })
  .listen(PORT).then(_ => {
    console.log(`> Running on localhost:${PORT}`);
  });
```

A simple method, like [in usage by express](https://expressjs.com/en/starter/static-files.html)
``` js
polka()
  .use(serve)
  .listen(PORT).then(_ => {
    console.log(`> Running on localhost:${PORT}`);
  });
```
is not possible currently, because `bwares` and `apps` require a base(for subapp) and the URL's are being stripped off in favour for subapps, so `serve-static` fails to handle the requests (I might be doing something wrong here, if any please help).

And as in subapps, the URL's are being stripped untill the first slash(`/`) to obtain the `base`. For requests like `/sw.js`, the `sw.js` is considered the `base`, and thus the requests are redirected to `/sw.js/`. This behaviour is expected but, generally(might be personally) it feels proper to have all the requests managed by the middleware. Simply being able to register middlewares to handle requests (not like compression, body-parser etc. though) is helpful.

**Comment by ramlmn** [comment]
Actually I was trying to make a point that `wares` and `bwares` might need to be called at different times (the loop() thingy called twice separately) while stripping the URL's for `bwares` and ones without for `wares`.

**Comment by lukeed** [comment]
Hey, thanks~! This was fixed in f4e62bf797c2599556d378d709b9911d64b1706d 🎉 And you can verify it with the changes in 1cfbb257ac23e880f58949998e080164e01b50a3.

If I understand you correctly, I think you were _mostly_ right! 😄 

I was mutating the `req.url` as soon as Polka couldn't find a main-app handler. It automatically assumed that there would be a sub-application/group that would handle the request. 

However, as you pointed out, this meant that something like `serve-static` wasn't getting all the information when it _wasn't_ mounted to a basepath. This means that something like `/app.js` was being sent to `serve-static` as `/`, which is definitely not what we wanted. As far as `serve-static` was concerned, we were requesting an `index.html`, and that's why that file was always sent, if it existed.

The fix is to _only_ mutate the request's `url` and `pathname` if we _know_, for certain, that a sub-something exists.

Thanks again~

**Comment by ramlmn** [comment]
A quick doubt. Wouldn't it be unfair to pass _normal_ URL's to the middewares when there are _no bwares_ at all and to mutate the URL's when there is _atleast one_ (which matches a base though) 🤔

**Comment by lukeed** [comment]
No, global middleware run as "the first line of defense" and should have the full URL at hand. The URL is only mutated once _entering_ a `bware` or a sub-application, because at that point you've declared that `{this group}` is prefixed by `basepath`, in which case, the prefix becomes redundant and repetitive information.

If you want a middleware to run on all routes, or it requires no manipulation, then it's (by def) a global middleware.

**Comment by lukeed** [comment]
It's the other way around. 

Your `serve1` sees the full URL: `/assets/foo/bar.png`

Your `serve2` receives a stripped URL: `/foo.bar.png` --- BUT it only sees this this if `serve1` passes along the request. 

In the case of `serve-static` (which is probably what you're referencing), this will only happen if `serve1` couldn't find an asset that matched `/assets/foo/bar.png` _starting from_ the directory you've initialized `serve1` with. 

So, for example, given this structure:

```
public
  |- assets
    |- foo
      |- bar.png
```

... and this code:

```js
const statics = require('serve-static');

const serve1 = statics('public');
const serve2 = statics('public/assets');

polka()
  .use(serve1) // <-- fixed, should be use()
  .use('assets', serve2)
```

An incoming request of `/assets/foo/bar.png` will **never** hit `serve2` because `serve1` found the file first. This is because of your file structure (and repetitive, unnecessary `serve2` declaration).

However, an incoming request of `/assets/foo/dog.jpg` will go thru `serve1` first. It will not find a matching file within `public/assets/foo`, so then, internally, it calls `next()`, which lets `serve2` try, but it will also fail since the file does not exist.

---

Clearly, this can be confusing. That is why `serve-static` is typically used once per app -- especially being that it's a massive, expensive middleware. It also carries its own "routing" logic.

The only way that multiple `serve-static` makes sense is with sibling directories; eg:

```
public
  |- foo
    |- foo.jpg
  |- bar
    |- bar.jpg
```

Then you can do `statics('public/foo')` and `statics('public/bar')` and mount those with basepaths. But, arguably, you should just do `statics('public')` since that's effectively the same thing.

**Comment by ramlmn** [comment]
```
// Directory structure is
public
  ├ assets
  │  ├ css
  │  │  └ style.css
  │  └ js
  │     └ app.js
  └ html
     └ index.html
```
``` html
<!-- public/html/index.html -->
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Example: Serve Static</title>
    <link rel="stylesheet" href="/assets/css/style.css">
  </head>
  <body>
    <div id="app" />
    <script src="https://unpkg.com/onloaded@1.0.0/dist/onloaded.min.js"></script>
    <script src="/assets/js/app.js"></script>
  </body>
</html>
```
``` js
// Consider this code
const serve = require('serve-static');
const serveAssets = serve('public/assets'); //  ┬- Two different directories
const serveHTML = serve('public/html');     //  ┘

polka()
  .use((req, res, next) => {
    // #globalware
    console.log(`Globalware req: ${req.url}`);
    serveHTML(req, res, next);  // Serves only HTML files
  })
  .use('assets', (req, res, next) => {
    // #baseware
    console.log(`Baseware req: ${req.url}`);
    serveAssets(req, res, next); // Handles all the assets
   })
  .listen(PORT).then(_ => {
    console.log(`> Running on localhost:${PORT}`);
  });
```
```
# My log is like

> Running on localhost:3000
Globalware req: /
Globalware req: /css/style.css
Baseware req: /css/style.css
Globalware req: /js/app.js
Baseware req: /js/app.js
```

The route `/` is handled by the `#globalware`

The log by `#globalware` should be
 * `/assests/css/style.css` instead of `/css/style.css` and
 * instead of `/js/app.js` it should be `/assets/js/app.js`

because it's a global middleware, right? Or am I doing something wrong (and, yes, the context is around `serve-static`)

**Comment by lukeed** [comment]
In this case, yes, the global middleware is mutating your `req.url` before the baseware sees the request. However, this is design & it's happening because your 2x `serve-static` is useless/redundant in this case.

This is happening because no routing `obj` definition is found, which means that the `app.handler` [looks for baseware](https://github.com/lukeed/polka/blob/master/packages/polka/index.js#L68). The calculated `base` is `/assets`, which matches a `bwares` key (because `.use('assets', ...)` mount).

If you changed your `.use('assets')` to `.use('statics')` (or anything other than "assets"), then the log will be as you expect. This is because the global middleware will still parse `assets` as a basepath, but not find any sub-groups with that key.

Worth mentioning, your setup is the same as my last comment: 

> But, arguably, you should just do statics('public') since that's effectively the same thing.

You are mounting two middleware that are splitting & navigating a structure that already exists. Consider `serve-static('public')` instead as a global middleware, if you must.

---

That said, you should _really_ consider **not** using `serve-static` as a global middleware. Doing so essentially uses `serve-static` as your router, rendering Polka useless to you. It also becomes your biggest bottleneck in performance since it has to traverse the file system for ***every*** request, determining file existence, before passing off the request to the rest of your application.

It's best used as an `assets` handler _only_, mounted to a base path _only_. This frees up the rest of your application, only hitting the super-slow-super-expensive middleware when directed to (`/assets/*`).

I recently made this change in an application of my own & saw a `req/s` increase of over 5,000!!!

**Comment by ramlmn** [comment]
Confessions:
* `serve-static` is only meant to be an example, I don't intend in using it anyway.
* my mounting points are `public/html` and `public/assets`, which are similar to `public/foo` and `public/bar`, not `public/` and `public/assets`

I'am mounting a middleware without a base path as its functionality is closer to a logging or monitoring tool. I _donot_ expect it to get stripped URL's regardless of any basepoints being mounted. That is the main reason for doing so.

I'am ignoring the perspective of performace because I am not going to use it for serving files. I just want to clear the routing part.

**Comment by lukeed** [comment]
Curious... Instead of:

```js
if (this.bwares[base] !== void 0) {
  mutate(req, info, base); //=> updates
  arr = arr.concat(this.bwares[base]);
}
```

Can you try this:

```js
if (this.bwares[base] !== void 0) {
  arr = arr.concat((req, res, next) => (mutate(req, info, base),next()), this.bwares[base]);
}
```

**Comment by ramlmn** [comment]
YES!!!. That is the code I was expecting to write. I wanted the mutation to occur between the end of calling wares and before the calling of bwares. This code is all I wanted to have. I have been to trying but failing to write this code. You have just written the mutation inside as a middleware(awesome!!!).

I always wanted to make this point, and finally you got it.

**Comment by lukeed** [comment]
Helps that you _actually_ told me what you were doing, instead of kinda-useless demos that implied behaviors that didn't actually have. 😆 

I'll look into benchmarking this & seeing that nothing inadvertently breaks.

Thanks~!

**Comment by ramlmn** [comment]
Something like
``` js
function mutateWare(req, res, next) {
  mutate(req, info, base);
  return next();
}

// ...

let base = value(info.pathname);
if (this.apps[base] !== void 0) {
  arr = arr.concat(mutateWare);
  fn = this.apps[base].handler.bind(null, req, res, info);
} else {
  fn = this.onNoMatch;
  if (this.bwares[base] !== void 0) {
    arr = arr.concat(mutateWare, this.bwares[base]);
  }
}
```
... is what I have in my mind.

**Comment by lukeed** [comment]
Nah, because you _do_ want to immediately mutate the request before passing it off to a sub-app.

**Comment by ramlmn** [comment]
Right, I need not mutate for sub-apps. Sorry.

**Comment by ramlmn** [comment]
Iam actually trying to rewrite my [serv](https://github.com/ramlmn/serv) package in polka ('cuz its fast and lightweight). serv is a small hacky dev server that I personally use for sharing files, and instantly serving static pages. I wanted to have a logger, which is a global middleware. It failed to work due to this issue, so Iam working on it. I'am rewring the whole thing to be small and most importantly fast.

**Comment by lukeed** [comment]
Cool~! I'll come back here once I've made the change. Not able to do it now, but it'll be out in the next patch 👍

**Comment by lukeed** [comment]
Also, if you're targeting Node 8+ (I think you are, since you're using `util.promisify`) you can use the native `http2` module instead of `spdy`.

**Comment by ramlmn** [comment]
Yes, Iam going to use native `http2` even though its still experimental API 😈. All my projects are hacky and for personal use only anyway.

**Comment by lukeed** [comment]
Looks promising 👍

**Comment by ramlmn** [comment]
😇😇😇

**Comment by lukeed** [comment]
This is the commit with the fix: 3120116882047c76040607b5c2984c6bb2c378ca

It wasn't quite as easy as the snippet above, but it works now. I also included tests & a [new example](https://github.com/lukeed/polka/tree/master/examples/with-morgan) to illustrate the changes.

Thanks again 🙌
